File: creator-project/assets/Script/ShipAI.js
cc.Class({
    extends: cc.Component,

    properties: {
        // foo: {
        //    default: null,
        //    url: cc.Texture2D,  // optional, default is typeof default
        //    serializable: true, // optional, default is true
        //    visible: true,      // optional, default is true
        //    displayName: 'Foo', // optional
        //    readonly: false,    // optional, default is false
        // },
        // ...
        bulletPrefab: {
            default: null,
            type: cc.Prefab
        }
    },

    // use this for initialization
    onLoad: function () {
        this._lastFire = Math.random() * 2;
        this._fireInterval = Math.random() * 1.7 + 0.8;
        this._speed = Math.random() * 25 + 0.5
        if (Math.random() >= 0.5) {
            this._speed = -this._speed;
        }
    },

    // called every frame, uncomment this function to activate update callback
    update: function (dt) {
        var node = this.node;
        var pos = node.getPosition();
        pos.x += this._speed * dt;
        node.setPosition(pos);
    
        if ((this._speed < 0 && pos.x < -400) || (this._speed > 0 && pos.x > 400)) {
            this._speed = -this._speed;
        }
    
        if (this._lastFire <= 0) {
            this._fireInterval = Math.random() * 1.7 + 0.8;
            this._lastFire = this._fireInterval;
    
            var bullet = cc.instantiate(this.bulletPrefab);
            var ai = bullet.getComponent("CannonBall01AI");
            ai.hittime = Math.random() * 0.6 + 0.75;
            ai.hitx = pos.x + Math.random() * 400 - 200;
            ai.hity = pos.y - Math.random() * 150 + 50;

            bullet.setPosition(pos);
            bullet.setLocalZOrder(1000);
            node.getParent().addChild(bullet);
        } else {
            this._lastFire = this._lastFire - dt;
        }
    },
});

File: lua-project/src/logic/ShipAI.lua

local creator = cc.import("creator.init")

local CannonBall01AI = cc.import(".CannonBall01AI")

local ShipAI = cc.class("ShipAI", creator.ComponentBase)

function ShipAI:onLoad(target)
    self._lastFire = math.random(0, 20) / 10
    self._fireInterval = math.random(8, 25) / 10
    self._speed = math.random(50, 300) / 10
    if math.random(1, 100) % 2 == 0 then
        self._speed = -self._speed
    end
end

function ShipAI:update(target, dt)
    local x, y = target:getPosition()
    x = x + self._speed * dt
    target:setPositionX(x)

    if (self._speed < 0 and x < -400) or (self._speed > 0 and x > 400) then
        self._speed = -self._speed
    end

    if self._lastFire <= 0 then
        self._fireInterval = math.random(8, 25) / 10
        self._lastFire = self._fireInterval

        local assets = creator.getAssets()
        local bullet = assets:createPrefab("resources/CannonBall01")
        bullet.hittime = math.random(75, 135) / 100
        bullet.hitx = x + math.random(-200, 200)
        bullet.hity = y - math.random(50, 200)
        bullet:addComponent(CannonBall01AI.new())
        bullet:trackComponents()

        bullet:setPosition(x, y)
        bullet:setLocalZOrder(1000)
        target:getParent():addChild(bullet)
    else
        self._lastFire = self._lastFire - dt
    end
end

return ShipAI
